curp/src/rpc/connect.rs

```diff
@@ -220,6 +237,98 @@ impl ConnectApi for Connect {
         req.set_timeout(timeout);
         client.fetch_leader(req).await.map_err(Into::into)
     }
+
+    async fn install_snapshot(
+        &self,
+        term: u64,
+        leader_id: ServerId,
+        snapshot: Snapshot,
+    ) -> Result<tonic::Response<InstallSnapshotResponse>, ProposeError> {
+        self.filter()?;
+
+        let mut client = self.get().await?;
+        client
+            .install_snapshot(Request::new(install_snapshot_stream(
+                term, leader_id, snapshot,
+            )))
+            .await
+            .map_err(Into::into)
+    }
+}
+
+/// Generate install snapshot stream
+fn install_snapshot_stream(
+    term: u64,
+    leader_id: ServerId,
+    snapshot: Snapshot,
+) -> impl Stream<Item = InstallSnapshotRequest> {
+    // FIXME: The following code is better. But it will result in an unknown compiling error that might origin from a compiler bug(https://github.com/rust-lang/rust/issues/102211).
+    // let req_stream = futures::stream::unfold(
+    //     (0, snapshot, term, leader_id, meta),
+    //     |(offset, mut snapshot, term, leader_id, meta)| async move {
+    //         if offset >= snapshot.size() {
+    //             return None;
+    //         }
+    //         let mut data = Vec::with_capacity(
+    //             std::cmp::min(snapshot.size() - offset, SNAPSHOT_CHUNK_SIZE).numeric_cast(),
+    //         );
+    //         let n: u64 = match snapshot.read(&mut data) {
+    //             Ok(n) => n.numeric_cast(),
+    //             Err(e) => {
+    //                 error!("read snapshot error, {e}");
+    //                 return None;
+    //             }
+    //         };
+    //         let done = (offset + n) == snapshot.size();
+    //         let req = InstallSnapshotRequest {
+    //             term,
+    //             leader_id: leader_id.clone(),
+    //             last_included_index: meta.last_included_index,
+    //             last_included_term: meta.last_included_term,
+    //             offset,
+    //             data,
+    //             done,
+    //         };
+    //         Some((req, (offset + n, snapshot, term, leader_id, meta)))
+    //     },
+    // );
+
+    let (tx, rx) = mpsc::channel(1);
+    let _ig = tokio::spawn(async move {
+        let meta = snapshot.meta;
+        let mut snapshot = snapshot.into_inner();
+        let mut offset = 0;
+        if let Err(e) = snapshot.rewind() {
+            error!("snapshot seek failed, {e}");
+            return;
+        }
+        #[allow(clippy::integer_arithmetic)] // can't overflow
+        while offset < snapshot.size() {
+            let len: u64 =
+                std::cmp::min(snapshot.size() - offset, SNAPSHOT_CHUNK_SIZE).numeric_cast();
+            let mut data = vec![0; len.numeric_cast()];
```

Comment:
Is it necessary to initialize the value to 0? Can we use the uninitialized vector for better perf?

Reply:
It seems that Rust is not ok with uninitialized memory in safe code.